### PR TITLE
#2742 - Attachment point selection edition and deletion

### LINKED ISSUE #2742: Attachment point selection edition and deletion
### Background
Based on inputs from @AlexanderSavelyev.
Currently, we have the tool 'Attachment point', and each time when we need to remove an attachment point or edit it we have to use this tool. The attachment point on Canvas is not hoverable and is not selectable.

### Proposal
We have to introduce new behavior for editing and removing attachment points.
1) Attachment point should be hoverable
2) There should be the ability to select an attachment point (with a selection tool) and see the selected attachment point.
3) Should be the ability to remove attachment points (select and `Del` hotkey or with Erase tool)
4) Should be the ability to edit an attachment point by clicking on the Context menu item 'Edit' -> dialog with attachment point type should be displayed (Choosing Primary/Secondary or both).

_Originally posted by @PetrKurkin in https://github.com/epam/ketcher/issues/2165#issuecomment-1434393564_

### Additional acceptance criteria (in the scope of the task)
We have to fix these issues (it is supposed to be fixed as we will have Attachment points as interaction elements on Canvas):
> https://github.com/epam/ketcher/assets/27288153/e7925166-e3b9-4ff1-aa35-0752c534946e 
> 1. After clicking _Save To Templates_, the attachment point touches the bounding. (wrong paddings)
> 2. When opening the Templates Modal to check out the saved template, the attachment point doesn't display in the preview

_Originally posted by @yuleicul_

### Additional context
Connected #2165

### PR DESCRIPTION
**Pull request by yuleicul**
Closes #2742 

This PR has been divided into 5 small PRs in order to ease code reviews.
This PR can be merged after all 5 PRs are approved, and then the 5 small PRs will be closed.

Part 1: Refactor #2928 
Part 2: Selection #2954
Part 3: Deletion #2960
Part 4: Bug fix #2969   
Part 5: Edition #2971 

## Check list
- [ ] unit-tests written
- [ ] e2e-tests written
- [ ] documentation updated
- [x] PR name follows the pattern `#1234 – issue name`
- [x] branch name doesn't contain '#'
- [x] PR is linked with the issue
- [x] base branch (master or release/xx) is correct
- [x] task status changed to "Code review"

### CODE REVIEW

File: packages/ketcher-core/src/application/editor/actions/rgroupAttachmentPoint.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+) {
+  const action = new Action();
+  action.mergeWith(fromRGroupAttachmentPointsDeletionByAtom(restruct, atomId));
+  action.mergeWith(fromRGroupAttachmentPointAddition(restruct, attpnt, atomId));
+  return action;
+}
+
+function fromRGroupAttachmentPointAddition(
+  restruct: ReStruct,
+  attpnt,
```
Comment (rrodionov91) [inline packages/ketcher-core/src/application/editor/actions/rgroupAttachmentPoint.ts:24]: Please use the full name for variables/parameters. Shorten ones can confuse newcommers in future

File: packages/ketcher-core/src/application/editor/actions/rgroupAttachmentPoint.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+  const actionToDeletePoint = new Action();
+  actionToDeletePoint.addOp(new RGroupAttachmentPointRemove(id));
+
+  return actionToDeletePoint
+    .perform(restruct)
+    .mergeWith(actionToUpdateAtomsAttrs);
+}
+
+function getNewAtomAttpnt(
```
Comment (rrodionov91) [inline packages/ketcher-core/src/application/editor/actions/rgroupAttachmentPoint.ts:101]: Here also

File: packages/ketcher-core/src/application/editor/actions/rgroupAttachmentPoint.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    action.addOp(new RGroupAttachmentPointRemove(rgroupAttachmentPointId));
+  });
+  return action.perform(restruct);
+}
+
+export function fromRGroupAttachmentPointDeletion(
+  restruct: ReStruct,
+  id: number,
+) {
+  const { atomId, newAttpnt } = getNewAtomAttpnt(restruct.molecule, id);
```
Comment (rrodionov91) [inline packages/ketcher-core/src/application/editor/actions/rgroupAttachmentPoint.ts:85]: I didn't catch why we need to get new attachment point during the deletion. Could you please explain?

File: packages/ketcher-core/src/application/editor/actions/rgroupAttachmentPoint.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  if (currentAttpnt === 1 || currentAttpnt === 2) {
+    newAttpnt = 0;
+  } else if (currentAttpnt === 3) {
+    const pointToDeleteType = pointToDelete.type;
+    if (pointToDeleteType === 'primary') {
+      newAttpnt = 2;
+    } else if (pointToDeleteType === 'secondary') {
+      newAttpnt = 1;
+    }
+  }
```
Comment (rrodionov91) [inline packages/ketcher-core/src/application/editor/actions/rgroupAttachmentPoint.ts:123]: I believe it's better to create constants for 0,1,2 and 1,2,3

Comment (rrodionov91) [reply packages/ketcher-core/src/application/editor/actions/rgroupAttachmentPoint.ts:123]: With meaningful name

File: packages/ketcher-core/src/application/editor/operations/OperationType.ts [hunk 5]
```diff
@@ -68,3 +70,24 @@ export const OperationType = Object.freeze({
...
+  R_GROUP_ATTACHMENT_POINT_ADD = 3,
+  S_GROUP_ATTR = 4,
+  ATOM_DELETE = 5,
+  FRAGMENT_STEREO_FLAG = 6,
+  CALC_IMPLICIT_H = 10,
+  FRAGMENT_DELETE_STEREO_ATOM = 90,
+  S_GROUP_DELETE = 95,
+  S_GROUP_ADD_TO_HIERACHY = 100,
+  FRAGMENT_ADD_STEREO_ATOM = 100,
+}
```
Comment (rrodionov91) [inline packages/ketcher-core/src/application/editor/operations/OperationType.ts:93]: Very good, thank you for that!

File: packages/ketcher-core/src/application/editor/operations/rgroupAttachmentPoint/RGroupAttachmentPointAdd.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    super(
+      OperationType.R_GROUP_ATTACHMENT_POINT_ADD,
+      OperationPriority.R_GROUP_ATTACHMENT_POINT_ADD,
+    );
+    this.data = data;
+  }
+
+  execute(restruct: ReStruct) {
+    const { atomId, attachmentPointType } = this.data;
+    const item = new RGroupAttachmentPoint(atomId, attachmentPointType);
```
Comment (rrodionov91) [inline packages/ketcher-core/src/application/editor/operations/rgroupAttachmentPoint/RGroupAttachmentPointAdd.ts:30]: I would suggest to rename 'item' to 'newAttachmentPoint' or something like that. Item is confusing a bit

File: packages/ketcher-core/src/application/render/restruct/reatom.ts [hunk 7]
```diff
@@ -564,6 +491,40 @@ class ReAtom extends ReObject {
...
+    }
+    da.push(angles[0] - angles[angles.length - 1] + 2 * Math.PI);
+    let daMax = 0;
+    let ang = -Math.PI / 2;
+    for (let i = 0; i < angles.length; ++i) {
+      if (da[i] > daMax) {
+        daMax = da[i];
+        ang = angles[i];
+      }
+    }
```
Comment (rrodionov91) [inline packages/ketcher-core/src/application/render/restruct/reatom.ts:517]: Please use full names for variables instead of 'hb', 'da', e.t.c

**Review by rrodionov91 — CHANGES_REQUESTED** [review]

**Comment by rrodionov91** [comment]
Also please check how it works with this change https://github.com/epam/ketcher/pull/2977 .

File: packages/ketcher-core/src/application/editor/actions/rgroupAttachmentPoint.ts (hunk 1 above)
Comment (yuleicul) [reply packages/ketcher-core/src/application/editor/actions/rgroupAttachmentPoint.ts:24]: `attpnt` is an attribute of atom, I'm not sure it was named after some specification or it was an abbreviation of `rgroupAttachmentPoint`, so in this situation, I think keeping the name `attpnt` can make developers understand it's the same as `attpnt` in entity atom. 
Something needs to do in the future: https://github.com/epam/ketcher/issues/2962

File: packages/ketcher-core/src/application/editor/actions/rgroupAttachmentPoint.ts (hunk 4 above)
Comment (yuleicul) [reply packages/ketcher-core/src/application/editor/actions/rgroupAttachmentPoint.ts:123]: Yes, you are right, but not in the scope of this PR, you can see:
https://github.com/epam/ketcher/issues/2962

File: packages/ketcher-core/src/application/editor/actions/rgroupAttachmentPoint.ts (hunk 3 above)
Comment (yuleicul) [reply packages/ketcher-core/src/application/editor/actions/rgroupAttachmentPoint.ts:85]: I've commented meanings of values of `attpnt` here https://github.com/epam/ketcher/pull/2961, will add type annotation to `attpnt` in https://github.com/epam/ketcher/issues/2962.

As you can see in https://github.com/epam/ketcher/pull/2961 `attpnt` has five values 
```typescript
/**
   *  Values can be `0 | 1 | 2 | 3 | null`.
   * `1` - has a `primary` R-Group Attachment Point
   * `2` - has a `secondary` R-Group Attachment Point
   * `3` - has a `primary` and a `secondary` R-Group Attachment Points
   * `null` and `0` both mean the atom has no R-Group Attachment Points
   */
```
Imagine you have an atom whose `attpnt` is `3`, which means it has both a `primary` and a `secondary` R-Group Attachment Points. Now you want to delete one of the attachment points. 
- If you delete `primary`, then `getNewAtomAttpnt` will give you a new attpnt `2`. 
- If you delete `secondary`, then `getNewAtomAttpnt` will give you a new attpnt `1`.

File: packages/ketcher-core/src/application/editor/operations/rgroupAttachmentPoint/RGroupAttachmentPointAdd.ts (hunk 6 above)
Comment (yuleicul) [reply packages/ketcher-core/src/application/editor/operations/rgroupAttachmentPoint/RGroupAttachmentPointAdd.ts]: Thank you. Fixed

File: packages/ketcher-core/src/application/render/restruct/reatom.ts (hunk 7 above)
Comment (yuleicul) [reply packages/ketcher-core/src/application/render/restruct/reatom.ts]: Fixed

**Comment by yuleicul** [comment]
@rrodionov91 really appreciate your detailed review!

**Review by rrodionov91 — APPROVED** [review]

**Comment by MartaWilliams** [comment]
Tests fail

ad. 3 - it's possible to remove the attachment point after selecting it and pressing Del hotkey or clicking the right mouse key, But after selecting the attachment and choosing Erase /scissors from the Top bar menu nothing is happening

ad.4 Editing existing attachments is not working - clicking the right mouse is giving only the "remove". According to AC should be possible to adit a type of attachment point (Choosing Primary/Secondary or both)

**Pull request merged**